docarray/array/array/io.py

```diff
@@ -177,37 +230,53 @@ def _write_bytes(
             elif protocol == 'pickle-array':
                 f.write(pickle.dumps(self))
             elif protocol in SINGLE_PROTOCOLS:
-                from rich import filesize
-
-                from docarray.utils.progress_bar import _get_progressbar
-
-                pbar, t = _get_progressbar(
-                    'Serializing', disable=not show_progress, total=len(self)
-                )
-
-                f.write(self._stream_header)
-
-                with pbar:
-                    _total_size = 0
-                    pbar.start_task(t)
-                    for doc in self:
-                        doc_bytes = doc.to_bytes(protocol=protocol, compress=compress)
-                        len_doc_as_bytes = len(doc_bytes).to_bytes(
-                            4, 'big', signed=False
-                        )
-                        all_bytes = len_doc_as_bytes + doc_bytes
-                        f.write(all_bytes)
-                        _total_size += len(all_bytes)
-                        pbar.update(
-                            t,
-                            advance=1,
-                            total_size=str(filesize.decimal(_total_size)),
+                f.write(
+                    b''.join(
+                        self.to_binary_stream(
+                            protocol=protocol,
+                            compress=compress,
+                            show_progress=show_progress,
                         )
+                    )
+                )
             else:
                 raise ValueError(
                     f'protocol={protocol} is not supported. Can be only {ALLOWED_PROTOCOLS}.'
                 )
 
+    def to_binary_stream(
+        self,
+        protocol: str = 'protobuf',
+        compress: Optional[str] = None,
+        show_progress: bool = False,
+    ) -> Iterator[bytes]:
+        from rich import filesize
+
+        from docarray.utils.progress_bar import _get_progressbar
+
+        pbar, t = _get_progressbar(
+            'Serializing', disable=not show_progress, total=len(self)
+        )
+
+        yield self._stream_header
+
+        with pbar:
+            _total_size = 0
+            pbar.start_task(t)
+            for doc in self:
+                doc_bytes = doc.to_bytes(protocol=protocol, compress=compress)
```

Comment:
This is the main bottleneck when deserializing s very large documents. Does this support streaming option?

Reply:
a single Document itself is a bottleneck?

Reply:
If its large enough and the docarray has a lot of documents then most of the doc bytes will stay in memory. Anyway its up to the user to decide how to load a huge docarray. I still think that we need to test the new version with the laion400m dataset available in the jina hub. Nobody investigated this [issue](https://github.com/docarray/docarray/issues/1121).